### ISSUE TITLE
question: wlr_cursor button events are not sent to toplevel?

### ISSUE THREAD

**Issue by Shinyzenith** [flacjacket/pywlroots#96]
Hi! I'm kind of puzzled by an issue that I'm facing, I apologize if this has nothing to do with pywlroots. If there's a better place where I can ask such questions, please do redirect me.

I'm [current writing a compositor](https://github.com/shinyzenith/nextwm) and I copied a lot from tinywl to get to a somewhat usable state and that includes the cursor setup.

I created a `self.cursor` object in my server class and attached the button event to it.
The event triggers too as per my logs without any errors.
However my focused toplevel doesn't get any click events? I have checked and made sure that the surface has focus (`surface.set_activated(True)`, `self.seat.keyboard_notify_enter(....)`)

**Comment by m-col** [comment]
There are a few independent state changes involved here that truly are independent:

- `surface.set_activated` tells the client whether or not it is activated, but that's all that does. It doesn't change any behaviour, it's primarily used by clients to draw themselves different, for example GTK windows fade-out a bit when they become 'deactivated' (whatever that means for the compositor)
- `seat.keyboard_notify_enter` this tells the seat where key events should go
- `seat.pointer_notify_enter` this is what you're missing, it tells the seat where pointer events should go

Note that these truly are independent - you could have 3 different clients that are activated, have keyboard focus, and have pointer focus respectively (if that made sense to the compositor)

**Comment by Shinyzenith** [comment]
>  you could have 3 different clients that are activated, have keyboard focus, and have pointer focus respectively (if that made sense to the compositor)

If I understand this correctly, does this mean 3 toplevels can receive keyboard events if the compositor allowed so?

**Comment by m-col** [comment]
No, the 3 bullet points above are 3 independent processes that do not necessarily have any interplay. Up to 1 surface can have pointer focus. Up to 1 surface can have keyboard focus (could be the same one, could be a different one). Any number of surfaces can be 'activated' because it isn't a functionality, it's just a message to the surface to tell it to _look_ activated.

**Comment by Shinyzenith** [comment]
> No, the 3 bullet points above are 3 independent processes that do not necessarily have any interplay. Up to 1 surface can have pointer focus. Up to 1 surface can have keyboard focus (could be the same one, could be a different one). Any number of surfaces can be 'activated' because it isn't a functionality, it's just a message to the surface to tell it to _look_ activated.

Got it. I have one last question to ask but it's offtopic for the thread, should I open another ticket or just ask here?

**Comment by m-col** [comment]
Easier to just ask here
